### PR TITLE
indexrec: add set operation index candidates

### LINKED ISSUE #73799: indexrec: add set operations to index candidates
We don't currently create index candidates for set operations, but we should do this. 

Consider this example:
```sql
EXPLAIN SELECT v FROM uniontest UNION SELECT k FROM uniontest
```
This query could use an index on `k` and an index on `v`, which would allow it to perform a streaming set operation.

### PR DESCRIPTION
**Pull request by nehageorge**
Previously, we did not have index candidates for set operations,
which led to missing or less-optimal index recommendations. This
commit adds index candidates for the output columns of UNION,
INTERSECT/ALL, and EXCEPT/ALL operations.

Fixes: #73799.

Release note: None

### CODE REVIEW

File: pkg/sql/opt/exec/execbuilder/testdata/union [hunk 1]
```diff
@@ -388,10 +402,6 @@ vectorized: true
       spans: FULL SCAN
 ·
 Diagram: https://cockroachdb.github.io/distsqlplan/decode.html#eJycUVFrnTAYfd-vCN9TSzN6Y98CA29XB4K9dkbGxhCJ5psTnHFJBOXifx_qtjtLu9L7FL6Tc3LOyXcE-7MBDsHnh2gfHsjFXShS8TGi5FOQ3MYiuCQiiIL3KSkokeRDEt8TWZDwkAaJmOF9FP1hjGQvZtown7-5w0ji5C5IyO0XUgCFVis8yB9ogX8FBhmFzugSrdVmho4LIVQD8B2Fuu16N8MZhVIbBH4EV7sGgUMqiwYTlArN9Q4oKHSybpZnh9EfxnzMh7xWA1AQnWwtJ2-BQtw7TnyP-gyyiYLu3cnBOlkhcDbR81KwbQpZ-LLIi1yelcJ7NsXJvG-1UWhQbYyzWfkS5Ykq92gqFOji7trbNknHDvl25UChwW_uwmdX1PeuLt-Zuvp-Gv92ZNT3nu1485qfTtB2urX4uOuTL-_mgqgqXD_M6t6U-GB0udisY7zoFkChdeuttw5hu1wtS_hXzF4hZo_F3n_FNxvxbsqmN78CAAD__zutJUU=
-·
```
Comment (rytaft) [inline pkg/sql/opt/exec/execbuilder/testdata/union:391]: Interesting that this recommendation disappeared

**Review by rytaft — APPROVED** [review]
LGTM

**Review by mgartner — APPROVED** [review]
:lgtm: 

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73811#-:-MqzshUb4cNIdIPYTDYJ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nehageorge)

File: pkg/sql/opt/exec/execbuilder/testdata/union (hunk 1 above)
Comment (nehageorge) [reply pkg/sql/opt/exec/execbuilder/testdata/union:391]: Yep, it's actually an improvement. The table already has an index on `(b, a)`, which would be better for this query. In https://github.com/cockroachdb/cockroach/pull/73346, we consider existing indexes, which should eliminate the possibility of a less optimal recommendation like this happening again.

**Comment by nehageorge** [comment]
TFTRs!

bors r+

**Pull request merged**